src/compiler/semantic/analyser.rs

```diff
@@ -243,6 +261,17 @@ impl Analyser {
             // State decl symbols are added in
             // add_state_decls
             Statement::StateDecl(_, _, _) => {}
+            Statement::Transition(dsym_ref, id, _) => {
+                // Find the corresponding symbol and add usage
+                if let Some(entry) = self.symbols.find_symbol(&self.cur_scope, id.name()) {
```

Comment:
@alxkzmn 
 
Wouldn't find_symbol return `None` if the symbol has not been declared before?

I think the best solution for the problem of usage before declaration is to add the usages as a second pass, after the symbol table has been created. That also has the advantege that if we don't need to use the usages in the compiler we don't have to run it. Also you could create it as a different function or struct from the semantic analyser. Perhaps `ProcessUsages` or something like that.

Reply:
Totally agree, right now it's not working for states that are declared later down the code. I'll refactor.

Reply:
- I see a problem with the "second pass" approach that stems out of our AST being non-uniform. What I mean by non-uniform is that the only way to traverse the tree is to know exactly what each level looks like. For example, the top-level declaration is a machine, and machine has a `block` of statements, and statements are all different types as well. For the "second pass" approach it means repeating a lot of code that descends the AST in the same way as the analyzer, through a series of custom `match` and analysis statements specific for each level of the tree. In such a case the language server code will have to repeat the custom tree traversal from the analyzer code.
- Another approach could be to make each level of AST implement a common trait like `Node` that has `children: Vec<Node>` during parsing. This way tree traversal becomes uniform regardless of the level - you can recursively go through children and look for any necessary information. This would allow the language server to be independent and not have to copy all the custom `match`-es (and maybe would enhance the analyzer as well, for example, by adding a `parent: Node` to each `Node` and then being able to look up and down the scope(s)).
- The hybrid approach in the middle of the spectrum is to do everything in the same place (record usages and other necessary info while analyzing) which is in this PR. Here I understand your concern that not all of the code would be necessary for the end goal of compiling the circuit.

Reply:
It would have been done with `Vec<Node>` in another language like Java. However, in rust `Vec<Node>` is not legal if `Node` is a trait. You could do `Vec<Box<dyn Node>>` but in my experience this has been very problematic what I have done it in rust. That is the reason it is implemented like this currently. Change it to  `Vec<Box<dyn Node>>` would be a lot of work.

For me what is problematic from your approach is that you are adding symbols in different places that were the original code was done and make the code more difficult to mantain because before there was only a place where a symbol was added.

However, now I remember that we found the same problem when building the semantic analyser. The only symbols that can be used before declaring them are the states, in the transitions. That is why we call `add_state_decls` before processing the content of the state.

Hence, the states symbols should be declared before being in any transition statement. The current code should work without adding symbols. I think, so it is even simpler in the end.

Or perhaps I am missing something here